storage/backup/backup-cli/src/coordinator/mod.rs

```diff
@@ -0,0 +1,301 @@
+                &rx,
+                Self::backup_epoch_endings,
+            )
+            .boxed_local();
+        let backup_state_snapshots = self
+            .backup_work_stream(
+                backup_state.latest_state_snapshot_version,
+                &rx,
+                Self::backup_state_snapshot,
+            )
+            .boxed_local();
+        let backup_transactions = self
+            .backup_work_stream(
+                backup_state.latest_transaction_version,
+                &rx,
+                Self::backup_transactions,
+            )
+            .boxed_local();
+
+        let mut all_work = stream::select_all(vec![
+            watch_db_state,
+            backup_epoch_endings,
+            backup_state_snapshots,
+            backup_transactions,
+        ]);
+
+        loop {
+            all_work
+                .next()
+                .await
+                .ok_or_else(|| anyhow!("Must be a bug: we never returned None."))?
+        }
+    }
+}
+
+impl BackupCoordinator {
+    async fn try_refresh_db_state(&self, db_state_broadcast: &watch::Sender<DbState>) {
+        match self.client.get_db_state().await {
+            Ok(s) => db_state_broadcast
+                .broadcast(s.expect("Db should have been bootstrapped."))
+                .map_err(|e| anyhow!("Receivers should not be cancelled: {}", e))
+                .unwrap(),
+            Err(e) => println!("Failed pulling DbState from local Libra node: {}", e),
+        };
+    }
+
+    async fn backup_epoch_endings(
+        &self,
+        last_epoch_ending_epoch_in_backup: Option<u64>,
+        db_state: DbState,
+    ) -> Result<Option<u64>> {
+        let (first, last) = get_batch_range(
+            last_epoch_ending_epoch_in_backup,
+            self.epoch_ending_batch_size,
+        );
+
+        // <= because `db_state.epoch` hasn't ended yet
+        if db_state.epoch <= last {
+            // wait for the next db_state update
+            return Ok(last_epoch_ending_epoch_in_backup);
+        }
+
+        EpochEndingBackupController::new(
+            EpochEndingBackupOpt {
+                start_epoch: first,
+                end_epoch: last + 1,
+            },
+            self.global_opt.clone(),
+            Arc::clone(&self.client),
+            Arc::clone(&self.storage),
+        )
+        .run()
+        .await?;
+
+        Ok(Some(last))
+    }
+
+    async fn backup_state_snapshot(
+        &self,
+        last_snapshot_version_in_backup: Option<Version>,
+        db_state: DbState,
+    ) -> Result<Option<Version>> {
+        // FIXME: do real calculation
+        let next_snapshot_version = get_next_snapshot(
+            last_snapshot_version_in_backup,
+            self.state_snapshot_interval,
+        );
+
+        if db_state.committed_version < next_snapshot_version {
+            // wait for the next db_state update
+            return Ok(last_snapshot_version_in_backup);
```

Comment:
why do we return a `Result<Option<_>>` here

Reply:
because the initial state may be None (meaning nothing in the backup), if that's what you meant